recommenders/datasets/mock/movielens.py

```diff
@@ -0,0 +1,140 @@
+    rating: Series[float] = Field(in_range={"min_value": 1, "max_value": 5})
+    timestamp: Series[str] = Field(eq="2022-2-22")
+    title: Series[str] = Field(eq="foo")
+    genre: Series[str] = Field(eq="genreA|0")
+
+    @classmethod
+    def get_df(
+        cls,
+        size: int = 3, seed: int = 100,
+        keep_first_n_cols: Optional[int] = None,
+        keep_title_col: bool = False, keep_genre_col: bool = False,
+    ) -> pandas.DataFrame:
+        """Return fake movielens dataset as a Pandas Dataframe with specified rows.
+
+        Args:
+            size (int): number of rows to generate
+            seed (int, optional): seeding the pseudo-number generation. Defaults to 100.
+            keep_first_n_cols (int, optional): keep the first n default movielens columns.
+            keep_title_col (bool): remove the title column if False. Defaults to True.
+            keep_genre_col (bool): remove the genre column if False. Defaults to True.
+
+        Returns:
+            pandas.DataFrame: a mock dataset
+        """
+        schema = cls.to_schema()
+        if keep_first_n_cols is not None:
+            if keep_first_n_cols < 1 or keep_first_n_cols > len(DEFAULT_HEADER):
+                raise ValueError(f"Invalid value for 'keep_first_n_cols': {keep_first_n_cols}. Valid range: [1-{len(DEFAULT_HEADER)}]")
+            schema = schema.remove_columns(DEFAULT_HEADER[keep_first_n_cols:])
+        if not keep_title_col:
+            schema = schema.remove_columns([DEFAULT_TITLE_COL])
+        if not keep_genre_col:
+            schema = schema.remove_columns([DEFAULT_GENRE_COL])
+
+        random.seed(seed)
+        # For more information on data synthesis, see https://pandera.readthedocs.io/en/latest/data_synthesis_strategies.html
+        return schema.example(size=size)
+
+    @classmethod
+    def get_spark_df(
+        cls,
+        spark,
+        size: int = 3, seed: int = 100,
+        keep_title_col: bool = False, keep_genre_col: bool = False,
+        tmp_path: Optional[str] = None,
+    ):
+        """Return fake movielens dataset as a Spark Dataframe with specified rows
+
+        Args:
+            spark (SparkSession): spark session to load the dataframe into
+            size (int): number of rows to generate
+            seed (int): seeding the pseudo-number generation. Defaults to 100.
+            keep_title_col (bool): remove the title column if False. Defaults to False.
+            keep_genre_col (bool): remove the genre column if False. Defaults to False.
+            tmp_path (str, optional): path to store files for serialization purpose
+                when transferring data from python to java.
+                If None, a temporal path is used instead
+
+        Returns:
+            pyspark.sql.DataFrame: a mock dataset
+        """
+        pandas_df = cls.get_df(size=size, seed=seed, keep_title_col=True, keep_genre_col=True)
+
+        # generate temp folder
```

Comment:
why are we writing the pandas dataframe to disk to read into spark dataframe?
you should be able to do something like `spark.createDataFrame(pandas_df, schema=schema)`

Reply:
Yes, I've tried exactly that (`spark.createDataFrame(pandas_df)`) and this actually incurs a very expensive communication between java and python each time the spark.dataframe is queried that actually becomes a new bottleneck. I think what happens is that when we create a spark.dataframe from pandas.dataframe, data remains in the memory as Python objects, and the spark.dataframe is merely a wrapper on top the Python object (thus the expensive java <-> python communication). To rid of this wrapper behavior, we need to serialized the data and reload into java. I think this is what pyarrow is trying to resolved.

This is at least what I observed in Spark 2.4.4 without [pyarrow ](https://docs.databricks.com/spark/latest/spark-sql/spark-pandas.html)enabled. Not sure if there's a better way to do this.

Reply:
can you enable pyarrow?
`spark.conf.set("spark.sql.execution.arrow.enabled", "true")`

Reply:
Let me try :).

Reply:
@gramhagen, here's the result of using pyarrow. Unfortunately, it didn't bring any performance improvements when compared to using explicit serialization.

Here's the runtime of a notebook WITHOUT any serialization/pyarrow enabled (vanilla `spark.createDataFrame(pandas_df, schema=schema)`):
![diversity_metric_opt_no_serialization](https://user-images.githubusercontent.com/19534002/135637322-1ba66a11-c23f-4b60-bfbd-3b9700795d76.PNG)

Here's the runtime with pyarrow (v0.14.0) enabled:
![diversity_metric_opt_no_serialization_w_pyarrow3_no_error](https://user-images.githubusercontent.com/19534002/135637582-a6e893ea-bec1-42f6-b3bd-f5f4d13e6767.PNG)

Here's the runtime with explicit serialization (i.e. writing a CSV to disk and `spark.read.csv`):
![diversity_metric_opt_serialization](https://user-images.githubusercontent.com/19534002/135637787-edef9cde-ee31-4780-bc74-52c6ae88005a.PNG)

I did [encounter this warning](https://www.py4u.net/discuss/1630531) when trying to enable pyarrow in spark 2, but I managed overcome it by downgrading to pyarrow==0.14.0, following the solution in the thread. Still it doesn't bring any improvements 🤣. I would we can we can revisit pyarrow after we migrate to spark 3.

Reply:
ok, makes sense. thanks for checking this.

Reply:
wow, that's a big difference. @laserprec do you think spark 3 will improve the situation?

Reply:
I hope so. Ultimately, this java-to-python inefficiency is what [pyarrow is designed to optimize for](https://spark.apache.org/docs/latest/api/python/user_guide/arrow_pandas.html#apache-arrow-in-pyspark). We could have some join operations in our notebooks that made it difficult to leverage pyarrow, but I think it's a good idea to revisit with Spark 3.

Reply:
I am not sure I follow all the stuff happening here, but Isn't there any simpler and faster way to create a mock Spark DF? A way that avoids going through Pandas DF and read / write to file. How about creating a random numpy array and initializing a Spark DF using that array? I suspect that would take less time, it would avoid the I/O overhead at least and probably numpy would be faster than pandas.

Reply:
@anargyri, that's a good point, thanks for bring that up. I think the main problem here is how we initialize the Spark DF. If we initialize from any Python object (whether it's pandas DF or numpy arrays), during code excecution, inevitably there will be a data exchange from python to Java because the data is in memory as some python object but spark runs on a JVM. This data exchange is source of this bottleneck, and I don't see a way around it as long as the data is a Python object. To your question, I think the easiest way to initialize a Spark DF would be reading from a CSV file. However, storing static data files in the repo will have its own challenge and inconvenience.

I find reusing existing libraries for fake data generation being a more flexible approach, so we don't have to reinvent the wheel. I find [pandera's schema](https://pandera.readthedocs.io/en/stable/schema_models.html#schema-models) definition by far the most intuitive and simple interface for fake data generation. @anargyri, if you have any other suggestions, I am open to other options too 😃.